### PR TITLE
fix 2251 refactor how micrometer integration is tested

### PR DESCRIPTION
**Pull request by ericbottard**
This PR reconsiders how micrometer integration is tested by flipping the logic:
- the main tests don't have micrometer on the classpath (and assert what happens then)
- a dedicated test-set (using the same plugin as what was used for Blockhound, albeit for different reasons) adds micrometer and tests what happens then.

### CODE REVIEW

File: reactor-core/build.gradle [hunk 1]
```diff
@@ -84,7 +74,7 @@ dependencies {
 	optional "org.slf4j:slf4j-api:$slf4jVersion"
 
 	//Optional Metrics
-	optional "io.micrometer:micrometer-core:$micrometerVersion"
+	compileOnly "io.micrometer:micrometer-core:$micrometerVersion"
```
Comment (bsideup) [inline reactor-core/build.gradle:77]: we should keep it as `optional`

**Comment by bsideup** [comment]
@ericbottard very nice idea! 👍

File: reactor-core/build.gradle (hunk 1 above)
Comment (ericbottard) [reply reactor-core/build.gradle:77]: Can you elaborate? Does it have to do with how this is handled when published in maven repos?

Comment (bsideup) [reply reactor-core/build.gradle:77]: `optional` goes to the generated POM and signals that _at least_ `$micrometerVersion` should be used, while `compileOnly` would be removed from the POM and there is no hint which version we used

Comment (ericbottard) [reply reactor-core/build.gradle:77]: Slight problem: the tests don't pass anymore :( 
Micrometer IS on the test classpath when using `optional` apparently...

Comment (ericbottard) [reply reactor-core/build.gradle:77]: Also, have a look at https://repo1.maven.org/maven2/io/projectreactor/reactor-core/3.3.7.RELEASE/reactor-core-3.3.7.RELEASE.pom, no particular mention of micrometer

Comment (ericbottard) [reply reactor-core/build.gradle:77]: Related question: why don't we put blockHound in the proper configuration created for the test-set? Would fix the compilation errors I'm seeing in IDEA as far as I understand.

Comment (simonbasle) [reply reactor-core/build.gradle:77]: ok so after discussing `optional` and looking at the pom, we'll track and kinda document the fact that optional dependencies are not in the pom anymore in #2262. the `optional` scope from our local copy of `OptionalDependenciesPlugin` is simply not made part of the pom, whereas the `compileOnly` vanilla scope results in `<scope>provided</scope>` in the pom, which we don't want.

File: reactor-core/build.gradle [hunk 2]
```diff
@@ -216,7 +197,7 @@ tasks.withType(Test).all {
 //always depend on testNoMicrometer, skip testNG on Travis, skip loops when not releasing
 //note that this way the tasks can be run individually
 check {
-	dependsOn testNoMicrometer
+	dependsOn withMicrometerTest
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/build.gradle:200]: I think not. Tests sets are often used for longer integration tests, and the default is to not incorporate them in the default lifecycle apparently:
https://github.com/unbroken-dome/gradle-testsets-plugin#running-tests-as-part-of-the-build

File: reactor-core/build.gradle (hunk 1 above)
Comment (ericbottard) [reply reactor-core/build.gradle:77]: >Related question: why don't we put blockHound in the proper configuration created for the test-set? Would fix the compilation errors I'm seeing in IDEA as far as I understand.

Created #2264

Comment (bsideup) [reply reactor-core/build.gradle:77]: Are you sure that `compileOnly` creates `<scope>provided</scope>`? It is not the same in other Gradle projects I've seen

Comment (ericbottard) [reply reactor-core/build.gradle:65]: We just tested and indeed, it does not.

**Pull request merged**